### PR TITLE
Allow tooling authors to provide custom postcss factory through options

### PR DESCRIPTION
**Pull request by phated**
This PR contains a:

- [ ] **bugfix**
- [x] new **feature**
- [ ] **code refactor**
- [x] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

Ref #510 

I know @alexander-akait said it is out-of-scope, but I needed to write the code anyway. So I figured I'd make the code suitable to upstream and submit it anyway—under the assumption this will be closed.

I'll be publishing my fork as `@phated/postcss-loader` for our use in Storybook until we can make the 7.0 breaking release—at which time, we'll be able to require users to rely on peerDeps and switch back to this upstream project.

Cheers! 🚀

### CODE REVIEW

**Review by alexander-akait — COMMENTED** [review]
Can you provide example of usage? Do you want to use `postcss@7` or `postcss@8`? Why not validate it on boilerplate side? In theory I can add this as the migration options (and when we drop postcss@7 support we remove it)

**Comment by phated** [comment]
Currently, the way a user consumes Storybook looks (more-or-less) like:

```
/users-project-root
├─┬ node_modules
│ └─┬ storybook
│   ├── postcss-loader
│   └── postcss@^7
└── storybook-config-file
```

The internally-included `postcss@^7` dependency cannot be removed before the 7.0 released, and users are using it implicitly.

However, a new option can be exposed that allows users to specify a custom postcss factory that has priority over the implicit dependency using this feature. So they would do:

```
/users-project-root
├─┬ node_modules
│ ├─┬ storybook
│ │ ├── postcss-loader
│ │ └── postcss@^7
│ └── postcss@^8 <-- add postcss 8 as top-level dep
└── storybook-config-file <-- add require("postcss")
```

**Review by alexander-akait — COMMENTED** [review]
Okay, let's rename `customPostcss` to `implementation`, we have the same option in `sass-loader` (https://github.com/webpack-contrib/sass-loader/blob/master/src/options.json#L4)

**Comment by phated** [comment]
@alexander-akait great thought! I've updated to be more consistent with sass-loader. Do you have any recommendations for getting the snapshots to succeed consistently across node versions?

File: test/cjs.test.js [hunk 1]
```diff
@@ -1,3 +1,4 @@
+// eslint-disable-next-line import/no-named-as-default, import/no-named-as-default-member
```
Comment (alexander-akait) [inline test/cjs.test.js:1]: I think it is global eslint problem, we don't have this rules

File: src/index.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import { getOptions } from "loader-utils";
 import { validate } from "schema-utils";
 
-import postcss from "postcss";
+import ourPostcss from "postcss";
```
Comment (alexander-akait) [inline src/index.js:4]: Let's return to `postcss`

File: README.md [hunk 3]
```diff
@@ -593,6 +593,38 @@ module.exports = {
 };
 ```
 
+### `implementation`
+
+Type: `Function`
+Default: `undefined`
```
Comment (alexander-akait) [inline README.md:599]: `require('postcss')`

**Comment by alexander-akait** [comment]
Also can you accept CLA?

**Comment by phated** [comment]
I think the CLA app is bugged

Screen Shot 2021-01-20 at 10 35 00 AM

File: src/index.js (hunk 2 above)
Comment (phated) [reply src/index.js:4]: The problem I ran into was naming the local variable `postcss` too. I can change that and put this back to `postcss`

**Comment by phated** [comment]
@alexander-akait I addressed your comments and squashed the commits. I removed my package-lock updates and other things that made the tests pass locally for me, I think those should be fixed outside this feature.

**Review by alexander-akait — APPROVED** [review]

**Comment by alexander-akait** [comment]
Thanks

**Pull request merged**